pkg/kvstore/allocator.go

```diff
@@ -0,0 +1,539 @@
+	for _, fn := range opts {
+		fn(a)
+	}
+
+	a.startGC()
+
+	return a, nil
+}
+
+// WithEvents enables AllocatorEvent to be received on Allocator.Events
+func WithEvents() AllocatorOption {
+	return func(a *Allocator) { a.startWatch() }
+}
+
+// WithSuffix sets the suffix of the allocator to the specified value
+func WithSuffix(v string) AllocatorOption {
+	return func(a *Allocator) { a.suffix = v }
+}
+
+// WithMin sets the minimum identifier to be allocated
+func WithMin(id ID) AllocatorOption {
+	return func(a *Allocator) { a.min = id }
+}
+
+// WithMax sets the maximum identifier to be allocated
+func WithMax(id ID) AllocatorOption {
+	return func(a *Allocator) { a.max = id }
+}
+
+// Delete deletes an allocator and stops the garbage collector
+func (a *Allocator) Delete() {
+	a.stopGC <- true
+	a.stopWatch()
+}
+
+// DeleteAllKeys will delete all keys
+func (a *Allocator) DeleteAllKeys() {
+	Client.DeleteTree(a.idPrefix)
+	Client.DeleteTree(a.valuePrefix)
+	Client.DeleteTree(a.lockPrefix)
+}
+
+func invalidKey(key, prefix string, deleteInvalid bool) {
+	log.Warningf("kvstore: Found invalid key %s outside of prefix %s. Deleting...", key, prefix)
+
+	if deleteInvalid {
+		Delete(key)
+	}
+}
+
+func (a *Allocator) keyToID(key string, deleteInvalid bool) ID {
+	if !strings.HasPrefix(key, a.idPrefix) {
+		invalidKey(key, a.idPrefix, deleteInvalid)
+		return 0
+	}
+
+	id, _ := strconv.ParseUint(key[len(a.idPrefix):], 10, 64)
+	return ID(id)
+}
+
+// Naive ID allocation mechanism.
+// FIXME: This should consider a random ID in the future
+func (a *Allocator) selectAvailableID() (ID, string) {
+	for id := a.min; id <= a.max; id++ {
+		if _, ok := a.cache[id]; !ok {
+			return id, id.String()
+		}
+	}
+
+	return 0, ""
+}
+
+// RefillCache invalidates and refills the local cache. Afterwards,
+// Allocator.Cache will contain entries for all IDs to key mappings as found in
+// the KVstore.
+func (a *Allocator) RefillCache() error {
+	// fetch all /id/ keys
+	ids, err := ListPrefix(a.idPrefix)
+	if err != nil {
+		return err
+	}
+
+	newCache := IDMap{}
+	for k, v := range ids {
+		if id := a.keyToID(k, true); id != 0 {
+			key, err := a.keyType.PutKey(string(v))
+			if err == nil {
+				newCache[id] = key
+			}
+		}
+	}
+
+	a.mutex.Lock()
+	a.cache = newCache
+	if !a.disableCache {
+		a.validCache = true
+	}
+	a.mutex.Unlock()
+
+	return nil
+}
+
+func (a *Allocator) allocate(key AllocatorKey, k []byte) (ID, error) {
+	a.mutex.RLock()
+	if !a.validCache {
+		// unlock for the duration of the cache refresh
+		a.mutex.RUnlock()
+
+		if err := a.RefillCache(); err != nil {
+			return 0, err
+		}
+
+		a.mutex.RLock()
+	}
+
+	id, strID := a.selectAvailableID()
+	a.mutex.RUnlock()
+	if id == 0 {
+		return 0, fmt.Errorf("no more identifiers available")
+	}
+
+	trace("selected available key %d", id)
+
+	// create /id/<ID> and fail if it already exists
+	err := CreateOnly(a.idPrefix+strID, k, false)
+	if err == nil {
+		// Add it to local cache
+		a.mutex.Lock()
```

Comment:
Doing the select and update atomically will require to hold a lock throughout the kvstore operations which is something I want to avoid. It will significantly reduce the ability to regenerate endpoints in parallel.

Instead, it seems like we want to make `selectAvailableID()` smart enough to handle concurrent calls by returning a random choice or being aware of the last n IDs handed out which have not been registered yet.